core/domain/learner_progress_services.py

```diff
@@ -0,0 +1,415 @@
+    """Save an activities completed domain object as an
+    ActivitiesCompletedByLearnerModel entity in the datastore.
+
+    Args:
+        activities_completed: ActivitiesCompletedByLearner. The activities
+            completed object to be saved in the datastore.
+    """
+    activities_completed_model = user_models.ActivitiesCompletedByLearnerModel(
+        id=activities_completed.id,
+        completed_exploration_ids=(
+            activities_completed.completed_exploration_ids),
+        completed_collection_ids=activities_completed.completed_collection_ids)
+
+    activities_completed_model.put()
+
+
+def save_incomplete_activities(incomplete_activities):
+    """Save an incomplete activities domain object as an
+    IncompleteActivitiesModel entity in the datastore.
+
+    Args:
+        incomplete_activities: IncompleteActivities. The incomplete
+            activities domain object to be saved in the datastore.
+    """
+    incomplete_activities_model = user_models.IncompleteActivitiesModel(
+        id=incomplete_activities.id,
+        incomplete_exploration_ids=(
+            incomplete_activities.incomplete_exploration_ids),
+        incomplete_collection_ids=(
+            incomplete_activities.incomplete_collection_ids))
+
+    incomplete_activities_model.put()
+
+
+def save_last_playthrough_information(last_playthrough_information):
+    """Save an ExpUserLastPlaythrough domain object as an
+    ExpUserLastPlaythroughModel entity in the datastore.
+
+    Args:
+        last_playthrough_information: ExpUserLastPlaythrough. The last
+            playthrough information domain object to be saved in the datastore.
+    """
+    last_playthrough_information_model = (
+        user_models.ExpUserLastPlaythroughModel(
+            id=last_playthrough_information.id,
+            user_id=last_playthrough_information.user_id,
+            exploration_id=last_playthrough_information.exploration_id,
+            last_played_exp_version=(
+                last_playthrough_information.last_played_exp_version),
+            last_played_state_name=(
+                last_playthrough_information.last_played_state_name)))
+    last_playthrough_information_model.put()
+
+
+def mark_exploration_as_completed(user_id, exp_id):
+    """Adds the exploration id to the completed list of the user unless the
+    exploration has already been completed or has been created/edited by the
+    user. It is also removed from the incomplete list (if present).
+
+    Args:
+        user_id: str. The id of the user who has completed the exploration.
+        exp_id: str. The id of the completed exploration.
+    """
+    activities_completed_model = (
+        user_models.ActivitiesCompletedByLearnerModel.get(
+            user_id, strict=False))
+    if not activities_completed_model:
+        activities_completed_model = (
+            user_models.ActivitiesCompletedByLearnerModel(id=user_id))
+
+    subscribed_exploration_ids = (
```

Comment:
Perhaps add a comment above this line explaining the design rationale (don't want to see things that overlap with creator dashboard).

(In the future I can see us amending this to whitelist "featured" explorations, e.g. in the basic math collections.)